### PR TITLE
Use xkb_state_update_mask() with Wayland backend

### LINKED ISSUE #158: Handle/support xkb in wlroots
The compositor examples currently use [`xkb_state_update_key`](https://github.com/SirCmpwn/wlroots/blob/master/examples/shared.c#L39).
The [documentation of this function](https://xkbcommon.org/doc/current/group__state.html#gac554aa20743a621692c1a744a05e06ce) states that `xkb_state_update_mask` should be used when input comes from a master process, like a wayland compositor.

This requires to either detect the current backend in a user, oder handle (part of) xkb inside wlroots.

### PR DESCRIPTION
**Pull request by versusvoid**
Fix #158

### CODE REVIEW

File: backend/libinput/keyboard.c [hunk 1]
```diff
@@ -23,9 +23,17 @@ static void wlr_libinput_keyboard_destroy(struct wlr_keyboard *wlr_kb) {
 	libinput_device_unref(wlr_libinput_kb->libinput_dev);
 }
 
+void wlr_libinput_keyboard_update_state(struct wlr_keyboard *keyboard,
+		struct wlr_event_keyboard_key *event) {
+	uint32_t keycode = event->keycode + 8;
+	xkb_state_update_key(keyboard->xkb_state, keycode,
+		event->state == WLR_KEY_PRESSED ? XKB_KEY_DOWN : XKB_KEY_UP);
```
Comment (emersion) [inline backend/libinput/keyboard.c]: I don't like how xkb stuff is mixed with libinput stuff.

File: include/wlr/interfaces/wlr_keyboard.h [hunk 2]
```diff
@@ -7,11 +7,16 @@
 struct wlr_keyboard_impl {
 	void (*destroy)(struct wlr_keyboard *keyboard);
 	void (*led_update)(struct wlr_keyboard *keyboard, uint32_t leds);
+	void (*update_state)(struct wlr_keyboard *keyboard,
+			struct wlr_event_keyboard_key *event);
```
Comment (emersion) [inline include/wlr/interfaces/wlr_keyboard.h]: Maybe we could add a `bool sends_modifiers_update` field instead? Not sure if it's a good idea to put a `bool` in a `*_impl` struct.

**Comment by emersion** [comment]
Also, we need to call `wlr_keyboard_update_modifiers` from xwayland.

File: include/wlr/interfaces/wlr_keyboard.h (hunk 2 above)
Comment (versusvoid) [reply include/wlr/interfaces/wlr_keyboard.h]: Nah, I think better solution will be `source` field in `struct wlr_event_keyboard_key` and condition on it.

**Comment by versusvoid** [comment]
And why we need it? It does not call `wlr_keyboard_update_state()` now and works fine.

File: include/wlr/interfaces/wlr_keyboard.h (hunk 2 above)
Comment (emersion) [reply include/wlr/interfaces/wlr_keyboard.h]: You mean, a field which indicates if keyboard state has already been updated?

**Comment by emersion** [comment]
The reason why it's better to call `wlr_keyboard_update_modifiers` when we don't have ownership over the keyboard (i.e. when we're not the primary compositor, when we're an X/wayland client) is that we can miss some keys (e.g. catch <kbd>Ctrl</kbd> pressed but miss <kbd>Ctrl</kbd> released) and end up with a corrupted keyboard state.

**Comment by emersion** [comment]
Also, it may be better to rename `wlr_keyboard_update_state` to something else (`wlr_keyboard_send_key`?) as this function won't even always update the keyboard state anymore.

File: include/wlr/interfaces/wlr_keyboard.h (hunk 2 above)
Comment (versusvoid) [reply include/wlr/interfaces/wlr_keyboard.h]: Something like
```cpp
 void wlr_keyboard_update_state(struct wlr_keyboard *keyboard,
  		struct wlr_event_keyboard_key *event) {
	if (event->source == LIBINPUT) {
		uint32_t keycode = event->keycode + 8;
		xkb_state_update_key(keyboard->xkb_state, keycode,
			event->state == WLR_KEY_PRESSED ? XKB_KEY_DOWN : XKB_KEY_UP);
	}
```

Comment (emersion) [reply include/wlr/interfaces/wlr_keyboard.h]: libinput may or may not be the only backend that doesn't call `wlr_keyboard_update_modifiers`. I would prefer a more descriptive name, such as `bool update_state`. And as soon as the event will be forwarded to the compositor and the compositor doesn't care about this, maybe make this an additional argument to this function. But you're right, it's a better idea.

File: backend/libinput/keyboard.c (hunk 1 above)
Comment (SirCmpwn) [reply backend/libinput/keyboard.c]: I agree. This shouldn't be here.

File: include/wlr/interfaces/wlr_keyboard.h (hunk 2 above)
Comment (SirCmpwn) [reply include/wlr/interfaces/wlr_keyboard.h]: -1 to having wlr_keyboard test for which backed the event came from

**Comment by versusvoid** [comment]
By looking at keyboard event handling in weston's X11 backend it can be concluded that correct approach isn't as simple as calling `wlr_keyboard_update_modifiers`. Committed modification make it work for now, but unlikely is complete solution.

**Comment by SirCmpwn** [comment]
Looks good, thanks! If you're concerned about the X11 backend, file a ticket, but I don't personally really care about that backend.

**Pull request merged**